### ISSUE TITLE
update_where() function

### ISSUE THREAD

**Issue by jcvall** [ericmjl/pyjanitor#153]
During my data cleaning..I find myself wanting to correct a row of data based on other column conditions.  For example, I wanted to update column 'C' based on column 'A' and column 'B'.  Where column 'A' has United States for example, column 'B' has 'Pennsylvania' and column C might be NaN but I want it to be the average salary where column A and B are present in the row.  And.....I dont want to overwrite any other data.  The way I do it now is create a table of column 'A' , column 'B', and column 'C' and perform a left join with df.merge.  Is there a better way?  I would like to add this function to pyjanitor.  I know the R usage of case_when a lot more flexible, but not sure what to call this function.

**Comment by ericmjl** [comment]
@jcvall I think this would be a great thing to add! Starting with `case_when` could be a good starting point; in which R package does this function exist? It can even be a port of the R functionality, I'd say.

**Comment by ericmjl** [comment]
Or we can start small and build it up from what your use case is, and slowly generalize it over time.

**Comment by jcvall** [comment]
In R...this function allows you to have multiple if and else if statements which are vectorize. It is equivalent to SQL CASE WHEN statement. The case_when function is found in the dplyr package.  So my use doesn't follow perfectly here.

Here is an idea I can get started with:

def case_when(df, condition1 , target1, condition2, target2 ):
    x = condition1.update(target1)
    y = x.update(target2)
    z = pd.DataFrame(y)
    df = df.merge(z, how = 'left')
    return df

A simple example is as follows:
case_when(df, {'country': ['US'],'state': ['PA']}, {'Avg. Salary' : [57000]})
This reads....case_when country = US and state = PA update Avg. Salary to 57000. Then above by adding more conditions and targets it acts like multiple if statements.

**Comment by ericmjl** [comment]
@jcvall yes, I'd love to see a PR for this! Please go ahead, I'll be more than happy to review!

**Comment by szuckerman** [comment]
I really like this addition, however, one question I have is how one would deal with `OR` conditions.

For example, with `CASE WHEN` in SQL, one can do this (I used 'UK' since 'PA' is a subset of 'US'):

```sql
SELECT
    CASE WHEN country = 'UK' OR state = 'PA' THEN 57000 ELSE 0 END
FROM TABLE
```

the above example forces the logical operator to be `AND`:

```sql
SELECT
    CASE WHEN country = 'US' AND state = 'PA' THEN 57000 ELSE 0 END
FROM TABLE
```

I don't know what the answer is, or even if R's `case_when` function takes care of this situation, too. Is there a way to nest calls to accomplish this behavior?

**Comment by jcvall** [comment]
Sorry. My code above was a scribble of an idea at the time.  I actually changed the code when I was working on it last night because it had some issues, to your point I didn't the limitations of not having the and, or = , > < != operators .  My new proposed code is even simpler and probably only slightly better -
` 
def case_when(df,conditions , target_col , target_value ):
    df.loc[conditions,target_col] = target_value
    return df
`

**Comment by jcvall** [comment]
This is odd...I am having permission errors.  

Authentication failed.  You may not have permissions to access the repository when I try to publish the branch.  Thoughts?

**Comment by ericmjl** [comment]
Might it be that you cloned ericmjl/pyjanitor without forking first? 

I have an idea: could you paste the output of:

```
$ git remote -v
```

here?

**Comment by jcvall** [comment]
That was the issue.  I gave it another try this morning before I ran off to work. Thanks.

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by jcvall** [comment]
Thinking about changing this from case_when to update_where which is a more popular sql command that basically does the same thing.  Thoughts?

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by jcvall** [comment]
I have this function locally on my computer and it is working great. (Though I did change target_col: None and target_val: None).  I can fix that.

Here are a few things to keep in mind:

1. each conditions must be inside ( ) and you need the df with the column...for example  (df.state == "PA"); (df.state == "PA) | (df.state == "NJ")
2. The target_col is a string name of column...for example 'avg_salary'
3. The target_val can be either a value(text,int,float,ect.) or another column (df.median_salary).  If column is used it will use the value from that column where the conditions are met.

A great reason why I like using a column if needed,  say I have NAs and I want to use another column to update them.  I could use the following:

`df.update_where((df.state.isnull(), 'state',df.providence)`

This will find the null values in state and update them with another column 'providence'. Not a perfect example but you get the idea.

**Comment by zbarry** [comment]
Resolved via #156 it looks like.
